src/goto-instrument/contracts/cfg_info.h

```diff
@@ -0,0 +1,203 @@
+    else if(expr.id() == ID_member)
+    {
+      const typet &type = to_member_expr(expr).struct_op().type();
+      if(
+        type.id() == ID_struct || type.id() == ID_struct_tag ||
+        type.id() == ID_union || type.id() == ID_union_tag)
+      {
+        return is_local_composite_access(to_member_expr(expr).compound());
+      }
+      else
+      {
+        throw unsupported_operation_exceptiont(
+          "is_local_composite_access: unexpected assignment to member of '" +
+          type.id_string() + "'");
+      }
+    }
+    else if(expr.id() == ID_if)
+    {
+      return is_local_composite_access(to_if_expr(expr).true_case()) &&
+             is_local_composite_access(to_if_expr(expr).false_case());
+    }
+    else if(expr.id() == ID_typecast)
+    {
+      return is_local_composite_access(to_typecast_expr(expr).op());
+    }
+    else if(
+      expr.id() == ID_byte_extract_little_endian ||
+      expr.id() == ID_byte_extract_big_endian)
+    {
+      return is_local_composite_access(to_byte_extract_expr(expr).op());
+    }
+    else if(expr.id() == ID_complex_real)
+    {
+      return is_local_composite_access(to_complex_real_expr(expr).op());
+    }
+    else if(expr.id() == ID_complex_imag)
+    {
+      return is_local_composite_access(to_complex_imag_expr(expr).op());
+    }
+    else
+    {
+      // matches ID_address_of, ID_dereference, etc.
+      return false;
+    }
+  }
+};
+
+// For goto_functions
+class function_cfg_infot : public cfg_infot
+{
+public:
+  function_cfg_infot(const goto_functiont &_goto_function)
+    : dirty_analysis(_goto_function), locals(_goto_function)
+  {
+    parameters.insert(
+      _goto_function.parameter_identifiers.begin(),
+      _goto_function.parameter_identifiers.end());
+  }
+
+  /// Returns true iff `ident` is a local or parameter of the goto_function.
+  bool is_local(const irep_idt &ident) const override
+  {
+    return locals.is_local(ident) ||
+           (parameters.find(ident) != parameters.end());
+  }
+
+  /// Returns true iff the given `ident` is either not a goto_function local
+  /// or is a local that is dirty.
+  bool is_not_local_or_dirty_local(const irep_idt &ident) const override
+  {
+    if(is_local(ident))
+      return dirty_analysis.get_dirty_ids().find(ident) !=
+             dirty_analysis.get_dirty_ids().end();
+    else
+      return true;
+  }
+
+private:
+  const dirtyt dirty_analysis;
+  const localst locals;
+  std::set<irep_idt> parameters;
+};
+
+// For a loop in a goto_function
+class loop_cfg_infot : public cfg_infot
+{
+public:
+  loop_cfg_infot(goto_functiont &_goto_function, const loopt &loop)
+    : dirty_analysis(_goto_function)
+  {
+    goto_programt::targett it = _goto_function.body.instructions.begin();
+    goto_programt::targett end = _goto_function.body.instructions.end();
+    while(it != end)
+    {
+      if(it->is_decl() && loop.contains(it))
+        locals.insert(it->decl_symbol().get_identifier());
+
+      it++;
+    }
```

Comment:
We can iterate over only instructions in the loop. 
```suggestion
    for(const auto &t : loop)
    {
      if(t->is_decl())
        locals.insert(t->decl_symbol().get_identifier());
    }
```

Reply:
Thanks !